geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/utils/GeoMessageSerializer.scala

```diff
@@ -8,46 +8,58 @@
 
 package org.locationtech.geomesa.kafka.utils
 
-import java.nio.ByteBuffer
 import java.nio.charset.StandardCharsets
-import java.time.Instant
 
 import com.typesafe.scalalogging.LazyLogging
 import org.apache.kafka.clients.producer.Partitioner
 import org.apache.kafka.common.Cluster
 import org.locationtech.geomesa.features.kryo.KryoFeatureSerializer
-import org.locationtech.geomesa.features.kryo.impl.KryoFeatureDeserialization
 import org.locationtech.geomesa.kafka.utils.GeoMessage.{Change, Clear, Delete}
+import org.locationtech.geomesa.utils.index.ByteArrays
 import org.opengis.feature.simple.SimpleFeatureType
 
 import scala.util.Random
+import scala.util.control.NonFatal
 import scala.util.hashing.MurmurHash3
 
 /**
   * Serialized `GeoMessage`s
   *
-  * The following encoding is used:
+  * Current encoding (version 2), designed to work with kafka log compaction.
+  * See https://kafka.apache.org/10/documentation.html#design_compactionbasics
   *
-  * Key: version (1 byte) type (1 byte) timestamp (8 bytes)
+  * change:
+  *   key: 1 byte message version, n bytes feature id
+  *   value: n bytes for serialized feature (without id)
```

Comment:
It was added in 0.10: https://kafka.apache.org/0100/javadoc/org/apache/kafka/clients/consumer/ConsumerRecord.html
For 0.9, in regular use, they will just get the timestamp of when they read the message instead of when it was written, which should be fairly close (and maintain the same ordering). Going back and reading a topic from the beginning, they would lose the info. However, the timestamps are only used by the feature listeners anyway, so in practice I'm not sure it would affect much.

Reply:
One option would be to rip the timestamp out of the KafkaFeatureEvent...  The listen command uses it...  I don't know if any other client code does.

Reply:
i was trying not to change the public API, else we'd technically have to bump major versions